### Data Summary
Import the libraries, load dataset, print shape of data, data description

In [1]:
import pandas as pd
import unicodedata
import nltk
import re
import spacy

from bs4 import BeautifulSoup
from nltk.tokenize.toktok import ToktokTokenizer

raw_data = pd.read_csv('Tweets.csv')

In [2]:
raw_data.shape

(14640, 15)

In [3]:
raw_data.head()

tweet_id airline_sentiment  airline_sentiment_confidence  \
0  570306133677760513           neutral                        1.0000   
1  570301130888122368          positive                        0.3486   
2  570301083672813571           neutral                        0.6837   
3  570301031407624196          negative                        1.0000   
4  570300817074462722          negative                        1.0000   

  negativereason  negativereason_confidence         airline  \
0            NaN                        NaN  Virgin America   
1            NaN                     0.0000  Virgin America   
2            NaN                        NaN  Virgin America   
3     Bad Flight                     0.7033  Virgin America   
4     Can't Tell                     1.0000  Virgin America   

  airline_sentiment_gold        name negativereason_gold  retweet_count  \
0                    NaN     cairdin                 NaN              0   
1                    NaN    jnardino                 NaN              0   
2                    NaN  yvonnalynn                 NaN              0   
3                    NaN    jnardino                 NaN              0   
4                    NaN    jnardino                 NaN              0   

                                                text tweet_coord  \
0                @VirginAmerica What @dhepburn said.         NaN   
1  @VirginAmerica plus you've added commercials t...         NaN   
2  @VirginAmerica I didn't today... Must mean I n...         NaN   
3  @VirginAmerica it's really aggressive to blast...         NaN   
4  @VirginAmerica and it's a really big bad thing...         NaN   

               tweet_created tweet_location               user_timezone  
0  2015-02-24 11:35:52 -0800            NaN  Eastern Time (US & Canada)  
1  2015-02-24 11:15:59 -0800            NaN  Pacific Time (US & Canada)  
2  2015-02-24 11:15:48 -0800      Lets Play  Central Time (US & Canada)  
3  2015-02-24 11:15:36 -0800            NaN  Pacific Time (US & Canada)  
4  2015-02-24 11:14:45 -0800            NaN  Pacific Time (US & Canada)

In [4]:
raw_data.describe()

tweet_id  airline_sentiment_confidence  negativereason_confidence  \
count  1.464000e+04                  14640.000000               10522.000000   
mean   5.692184e+17                      0.900169                   0.638298   
std    7.791112e+14                      0.162830                   0.330440   
min    5.675883e+17                      0.335000                   0.000000   
25%    5.685592e+17                      0.692300                   0.360600   
50%    5.694779e+17                      1.000000                   0.670600   
75%    5.698905e+17                      1.000000                   1.000000   
max    5.703106e+17                      1.000000                   1.000000   

       retweet_count  
count   14640.000000  
mean        0.082650  
std         0.745778  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max        44.000000

In [5]:
print(pd.unique(raw_data['airline_sentiment']))
print('')
print(pd.unique(raw_data['negativereason']))
print('')
print(pd.unique(raw_data['airline']))

['neutral' 'positive' 'negative']

[nan 'Bad Flight' "Can't Tell" 'Late Flight' 'Customer Service Issue'
 'Flight Booking Problems' 'Lost Luggage' 'Flight Attendant Complaints'
 'Cancelled Flight' 'Damaged Luggage' 'longlines']

['Virgin America' 'United' 'Southwest' 'Delta' 'US Airways' 'American']


### Data Columns
1. Drop all columns except `text` and `airline_sentiment`
2. Print shape of data
3. Print first 5 rows of data

In [6]:
data = raw_data.drop(raw_data.columns.difference(['text','airline_sentiment']), 1)

In [7]:
data.shape

(14640, 2)

In [8]:
data.head(5)

airline_sentiment                                               text
0           neutral                @VirginAmerica What @dhepburn said.
1          positive  @VirginAmerica plus you've added commercials t...
2           neutral  @VirginAmerica I didn't today... Must mean I n...
3          negative  @VirginAmerica it's really aggressive to blast...
4          negative  @VirginAmerica and it's a really big bad thing...

### Data Pre-Processing
1. HTML tag removal
2. Tokenization
3. Remove numbers
4. Remove special characters and punctuation
5. Convert to lowercase
6. Lemmatize or stem
7. Join words into text string
8. Print first 5 rows after pre-processing

In [9]:
# HTML tag removal
def strip_html(text):
    soup = BeautifulSoup(text, "html.parser")
    return soup.get_text()

data['text'] = data['text'].apply(strip_html)

data.head()

airline_sentiment                                               text
0           neutral                @VirginAmerica What @dhepburn said.
1          positive  @VirginAmerica plus you've added commercials t...
2           neutral  @VirginAmerica I didn't today... Must mean I n...
3          negative  @VirginAmerica it's really aggressive to blast...
4          negative  @VirginAmerica and it's a really big bad thing...

In [10]:
# Remove numbers, special characters and punctuation
def remove_sc(text):
    pattern = r'[^A-Za-z\s]'
    return re.sub(pattern, '', text)

data['text'] = data['text'].apply(remove_sc)

data.head()

airline_sentiment                                               text
0           neutral                   VirginAmerica What dhepburn said
1          positive  VirginAmerica plus youve added commercials to ...
2           neutral  VirginAmerica I didnt today Must mean I need t...
3          negative  VirginAmerica its really aggressive to blast o...
4          negative  VirginAmerica and its a really big bad thing a...

In [11]:
# Convert to lowercase
def lower(text):
    return text.lower()

data['text'] = data['text'].apply(lower)

data.head()

airline_sentiment                                               text
0           neutral                   virginamerica what dhepburn said
1          positive  virginamerica plus youve added commercials to ...
2           neutral  virginamerica i didnt today must mean i need t...
3          negative  virginamerica its really aggressive to blast o...
4          negative  virginamerica and its a really big bad thing a...

In [12]:
# Tokenization
stopwords = nltk.corpus.stopwords.words('english')

def tokenize(text):
    tokenizer = ToktokTokenizer()
    tokens = tokenizer.tokenize(text)
    tokens = [token.strip() for token in tokens if token not in stopwords]
    return tokens

data['text'] = data['text'].apply(tokenize)

data.head()

airline_sentiment                                               text
0           neutral                    [virginamerica, dhepburn, said]
1          positive  [virginamerica, plus, youve, added, commercial...
2           neutral  [virginamerica, didnt, today, must, mean, need...
3          negative  [virginamerica, really, aggressive, blast, obn...
4          negative           [virginamerica, really, big, bad, thing]

In [13]:
# Join words into text string
def join(tokens):
    return ' '.join(tokens)

data['text'] = data['text'].apply(join)

data.head()

airline_sentiment                                               text
0           neutral                        virginamerica dhepburn said
1          positive  virginamerica plus youve added commercials exp...
2           neutral  virginamerica didnt today must mean need take ...
3          negative  virginamerica really aggressive blast obnoxiou...
4          negative                 virginamerica really big bad thing

In [14]:
# Lemmatize or stem
nlp = spacy.load('en_core_web_sm')

def lemma(text):
    text = nlp(text)
    return ' '.join([word.lemma_ if word.lemma_ != '-PRON-' else word.text for word in text])

data['text'] = data['text'].apply(lemma)

In [15]:
# Print first 5 rows after pre-processing
data.head()

airline_sentiment                                               text
0           neutral                         virginamerica dhepburn say
1          positive  virginamerica plus you ve add commercial exper...
2           neutral  virginamerica do nt today must mean need take ...
3          negative  virginamerica really aggressive blast obnoxiou...
4          negative                 virginamerica really big bad thing

In [16]:
# Replace sentiment text with numerical values
replaceStruct = {
    'airline_sentiment': {
        'positive': 1,
        'neutral': 0,
        'negative': -1
    }
}

data = data.replace(replaceStruct)

data.head()

airline_sentiment                                               text
0                  0                         virginamerica dhepburn say
1                  1  virginamerica plus you ve add commercial exper...
2                  0  virginamerica do nt today must mean need take ...
3                 -1  virginamerica really aggressive blast obnoxiou...
4                 -1                 virginamerica really big bad thing

In [17]:
# Split data into training and testing
from sklearn.model_selection import train_test_split

x = data.drop('airline_sentiment', axis=1)
y = data['airline_sentiment']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=.3, random_state=42)

### Vectorization
1. Use `CountVectorizer`
2. Use `TfidfVectorizer`

In [18]:
# CountVectorizer
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()

In [19]:
# TF-IDF Vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

tf = TfidfVectorizer()

### Modelling and Evaluation
Fit and evaluate model using both types of vectorization

In [20]:
train_cv = cv.fit_transform(data['text'])
train_tf = tf.fit_transform(data['text'])

In [21]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
import numpy as np

forest_cv = RandomForestClassifier(verbose=2, n_jobs=-1, n_estimators=100)
forest_cv = forest_cv.fit(train_cv, data['airline_sentiment'])

print(np.mean(cross_val_score(forest_cv, train_cv, data['airline_sentiment'], cv=10)))

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 4 concurrent workers.


building tree 1 of 100
building tree 2 of 100
building tree 3 of 100
building tree 4 of 100
building tree 5 of 100
building tree 6 of 100
building tree 7 of 100
building tree 8 of 100
building tree 9 of 100
building tree 10 of 100
building tree 11 of 100
building tree 12 of 100
building tree 13 of 100
building tree 14 of 100
building tree 15 of 100
building tree 16 of 100
building tree 17 of 100
building tree 18 of 100
building tree 19 of 100
building tree 20 of 100
building tree 21 of 100
building tree 22 of 100
building tree 23 of 100
building tree 24 of 100
building tree 25 of 100
building tree 26 of 100
building tree 27 of 100
building tree 28 of 100
building tree 29 of 100
building tree 30 of 100
building tree 31 of 100
building tree 32 of 100
building tree 33 of 100
building tree 34 of 100
building tree 35 of 100
building tree 36 of 100


[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    3.2s


building tree 37 of 100
building tree 38 of 100
building tree 39 of 100
building tree 40 of 100
building tree 41 of 100
building tree 42 of 100
building tree 43 of 100
building tree 44 of 100
building tree 45 of 100
building tree 46 of 100
building tree 47 of 100
building tree 48 of 100
building tree 49 of 100
building tree 50 of 100
building tree 51 of 100
building tree 52 of 100
building tree 53 of 100
building tree 54 of 100
building tree 55 of 100
building tree 56 of 100
building tree 57 of 100
building tree 58 of 100
building tree 59 of 100
building tree 60 of 100
building tree 61 of 100
building tree 62 of 100
building tree 63 of 100
building tree 64 of 100
building tree 65 of 100
building tree 66 of 100
building tree 67 of 100
building tree 68 of 100
building tree 69 of 100
building tree 70 of 100
building tree 71 of 100
building tree 72 of 100
building tree 73 of 100
building tree 74 of 100
building tree 75 of 100
building tree 76 of 100
building tree 77 of 100
building tree 78

[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:    9.9s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    8.5s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   19.8s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    5.6s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   15.8s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_j

0.7325136612021859


[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   14.7s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:    0.0s finished


In [22]:
forest_tf = RandomForestClassifier(verbose=2, n_jobs=-1, n_estimators=100)
forest_tf = forest_tf.fit(train_tf, data['airline_sentiment'])

print(np.mean(cross_val_score(forest_tf, train_tf, data['airline_sentiment'], cv=10)))

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 4 concurrent workers.


building tree 1 of 100building tree 2 of 100

building tree 3 of 100
building tree 4 of 100
building tree 5 of 100
building tree 6 of 100
building tree 7 of 100
building tree 8 of 100
building tree 9 of 100building tree 10 of 100

building tree 11 of 100
building tree 12 of 100
building tree 13 of 100
building tree 14 of 100
building tree 15 of 100
building tree 16 of 100
building tree 17 of 100
building tree 18 of 100
building tree 19 of 100
building tree 20 of 100
building tree 21 of 100
building tree 22 of 100
building tree 23 of 100
building tree 24 of 100
building tree 25 of 100
building tree 26 of 100
building tree 27 of 100
building tree 28 of 100
building tree 29 of 100
building tree 30 of 100
building tree 31 of 100
building tree 32 of 100
building tree 33 of 100
building tree 34 of 100
building tree 35 of 100
building tree 36 of 100


[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    3.2s


building tree 37 of 100
building tree 38 of 100
building tree 39 of 100
building tree 40 of 100
building tree 41 of 100
building tree 42 of 100
building tree 43 of 100
building tree 44 of 100
building tree 45 of 100
building tree 46 of 100
building tree 47 of 100
building tree 48 of 100
building tree 49 of 100
building tree 50 of 100
building tree 51 of 100
building tree 52 of 100
building tree 53 of 100
building tree 54 of 100
building tree 55 of 100
building tree 56 of 100
building tree 57 of 100
building tree 58 of 100
building tree 59 of 100
building tree 60 of 100
building tree 61 of 100
building tree 62 of 100
building tree 63 of 100
building tree 64 of 100
building tree 65 of 100
building tree 66 of 100
building tree 67 of 100
building tree 68 of 100
building tree 69 of 100
building tree 70 of 100
building tree 71 of 100
building tree 72 of 100
building tree 73 of 100
building tree 74 of 100
building tree 75 of 100
building tree 76 of 100
building tree 77 of 100
building tree 78

[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:    9.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    4.6s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   14.1s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed:    4.5s
[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   14.0s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:    0.0s finished
[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_j

0.7216530054644809


[Parallel(n_jobs=-1)]: Done 100 out of 100 | elapsed:   15.7s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  33 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:    0.0s finished


### Conclusion
Summarize your understanding of the application of various pre-processing and vectorization and performance of your model on this dataset

* HTML tag removal removes any HTML tags from the content (not relevant for this dataset)
* Tokenization divides the text into smaller parts as a base step prior to lemmatization or stemming
* We remove numbers, special characters, and punctuation in order to more accurately tokenization and lemma/stem the content
* Conversion to lowercase reduces irrelevant differences in text meaning
* `CountVectorizer` is a basic count of individual tokens within a string
* `TF-IDF` normalizes token count across all documents in the corpus, in order to avoid erroneously high importance of words common throughout a given dataset
* In the above dataset, `CountVectorizer` has an accuracy of 73.25% using an n=10 cross validation
* In the above dataset, `TfidfVectorizer` has an accuracy of 72.17% using an n=10 cross validation